### PR TITLE
`torch.linalg.householder_product` faster backward

### PR DESCRIPTION
**Pull request by nikitaved**
This PR implements a much more efficient algorithm. This algorithm allows to achieve MASSIVE speed-ups, especially for batched and/or larger double-precision inputs.
Here are some benchmarks:

<details>

<summary>Testing script</summary>

```python
from IPython import get_ipython
import torch
import itertools

torch.manual_seed(13)
#torch.set_num_threads(1)

ipython = get_ipython()

cpu = torch.device('cpu')
cuda = torch.device('cuda')

def generate_input(shape, dtype=torch.double, device=cpu):
    eigvals = torch.rand(*shape[:-1], dtype=dtype, device=device)
    eigvecs = torch.rand(*shape, dtype=dtype, device=device)
    input = (eigvecs * eigvals.unsqueeze(-2)) @ eigvecs.inverse()
    input.requires_grad_(True)
    tau = torch.rand(*shape[:-1], dtype=dtype, device=device)
    tau.requires_grad_(True)
    return input, tau


def run_test(shape, device, dtype):
    print(f"shape: {shape}, device: {device}, dtype: {dtype}")
    a, tau = generate_input(shape, dtype=dtype, device=device)
    prod = torch.linalg.householder_product(a, tau)
    ones_prod = torch.ones_like(prod)

    command = "torch.autograd.backward((prod,), (ones_prod), retain_graph=True)"
    if device == cuda:
        command = command + "; torch.cuda.synchronize()"
    ipython.magic(f"timeit {command}")
    print()

dtypes = [torch.float, torch.double]
devices = [cpu, cuda]
#devices = [cuda]
sizes = [
    (10, 10),
    (1000, 10, 10),
    (100, 100),
    (1000, 100, 100),
    (1000, 1000),
    (10, 1000, 1000),
]

for device, dtype, size in itertools.product(devices, dtypes, sizes):
    run_test(size, device, dtype)

```

</details>

<details>

<summary>This PR, cuda float32</summary>

```
shape: (10, 10), device: cuda, dtype: torch.float32
1.33 ms ± 1.82 µs per loop (mean ± std. dev. of 7 runs, 1000 loops each)

shape: (1000, 10, 10), device: cuda, dtype: torch.float32
1.52 ms ± 40.8 µs per loop (mean ± std. dev. of 7 runs, 1000 loops each)

shape: (100, 100), device: cuda, dtype: torch.float32
10.8 ms ± 9.62 µs per loop (mean ± std. dev. of 7 runs, 100 loops each)

shape: (1000, 100, 100), device: cuda, dtype: torch.float32
127 ms ± 8.45 µs per loop (mean ± std. dev. of 7 runs, 10 loops each)

shape: (1000, 1000), device: cuda, dtype: torch.float32
151 ms ± 127 µs per loop (mean ± std. dev. of 7 runs, 10 loops each)

shape: (10, 1000, 1000), device: cuda, dtype: torch.float32
981 ms ± 91.4 µs per loop (mean ± std. dev. of 7 runs, 1 loop each)

```

</details>

<details>

<summary>Master, cuda float32</summary>

```
shape: (10, 10), device: cuda, dtype: torch.float32
1.64 ms ± 6.36 µs per loop (mean ± std. dev. of 7 runs, 1000 loops each)

shape: (1000, 10, 10), device: cuda, dtype: torch.float32
298 ms ± 463 µs per loop (mean ± std. dev. of 7 runs, 1 loop each)

shape: (100, 100), device: cuda, dtype: torch.float32
15.4 ms ± 41.5 µs per loop (mean ± std. dev. of 7 runs, 100 loops each)

shape: (1000, 100, 100), device: cuda, dtype: torch.float32
5.36 s ± 711 µs per loop (mean ± std. dev. of 7 runs, 1 loop each)

shape: (1000, 1000), device: cuda, dtype: torch.float32
1.64 s ± 1.07 ms per loop (mean ± std. dev. of 7 runs, 1 loop each)

shape: (10, 1000, 1000), device: cuda, dtype: torch.float32
15.7 s ± 43.8 ms per loop (mean ± std. dev. of 7 runs, 1 loop each)

```

</details>

<details>

<summary>This PR, cuda float64</summary>

```
shape: (10, 10), device: cuda, dtype: torch.float64
1.14 ms ± 1.43 µs per loop (mean ± std. dev. of 7 runs, 1000 loops each)

shape: (1000, 10, 10), device: cuda, dtype: torch.float64
2.22 ms ± 1.32 µs per loop (mean ± std. dev. of 7 runs, 100 loops each)

shape: (100, 100), device: cuda, dtype: torch.float64
10.6 ms ± 11.6 µs per loop (mean ± std. dev. of 7 runs, 100 loops each)

shape: (1000, 100, 100), device: cuda, dtype: torch.float64
287 ms ± 84.9 µs per loop (mean ± std. dev. of 7 runs, 1 loop each)

shape: (1000, 1000), device: cuda, dtype: torch.float64
236 ms ± 41.9 µs per loop (mean ± std. dev. of 7 runs, 1 loop each)

shape: (10, 1000, 1000), device: cuda, dtype: torch.float64
1.88 s ± 88.3 µs per loop (mean ± std. dev. of 7 runs, 1 loop each)

```
</details>

<details>

<summary>Master, cuda float64</summary>

```
shape: (10, 10), device: cuda, dtype: torch.float64
1.58 ms ± 8.21 µs per loop (mean ± std. dev. of 7 runs, 1000 loops each)

shape: (1000, 10, 10), device: cuda, dtype: torch.float64
308 ms ± 213 µs per loop (mean ± std. dev. of 7 runs, 1 loop each)

shape: (100, 100), device: cuda, dtype: torch.float64
79 ms ± 14.5 µs per loop (mean ± std. dev. of 7 runs, 10 loops each)

shape: (1000, 100, 100), device: cuda, dtype: torch.float64
54.2 s ± 1.9 ms per loop (mean ± std. dev. of 7 runs, 1 loop each)

shape: (1000, 1000), device: cuda, dtype: torch.float64
31.5 s ± 698 µs per loop (mean ± std. dev. of 7 runs, 1 loop each)

shape: (10, 1000, 1000), device: cuda, dtype: torch.float64
4min 45s ± 2.48 s per loop (mean ± std. dev. of 7 runs, 1 loop each)

```
</details>

<details>

<summary>This PR, cpu float32</summary>

```
shape: (10, 10), device: cpu, dtype: torch.float32        
476 µs ± 21.4 µs per loop (mean ± std. dev. of 7 runs, 1 loop each)

shape: (1000, 10, 10), device: cpu, dtype: torch.float32
5.1 ms ± 100 µs per loop (mean ± std. dev. of 7 runs, 100 loops each)

shape: (100, 100), device: cpu, dtype: torch.float32      
4.38 ms ± 4.2 µs per loop (mean ± std. dev. of 7 runs, 100 loops each)

shape: (1000, 100, 100), device: cpu, dtype: torch.float32
1.55 s ± 6.64 ms per loop (mean ± std. dev. of 7 runs, 1 loop each)     

shape: (1000, 1000), device: cpu, dtype: torch.float32   
745 ms ± 407 µs per loop (mean ± std. dev. of 7 runs, 1 loop each)      

shape: (10, 1000, 1000), device: cpu, dtype: torch.float32
5.44 s ± 15.4 ms per loop (mean ± std. dev. of 7 runs, 1 loop each)    

```
</details>

<details>

<summary>Master, cpu float32</summary>

```
shape: (10, 10), device: cpu, dtype: torch.float32                       
387 µs ± 645 ns per loop (mean ± std. dev. of 7 runs, 1000 loops each)
                                                                         
shape: (1000, 10, 10), device: cpu, dtype: torch.float32
12.3 ms ± 23.3 µs per loop (mean ± std. dev. of 7 runs, 100 loops each)
                                                                                        
shape: (100, 100), device: cpu, dtype: torch.float32      
39.4 ms ± 80.3 µs per loop (mean ± std. dev. of 7 runs, 10 loops each)
                                                                             
shape: (1000, 100, 100), device: cpu, dtype: torch.float32                 
29.1 s ± 44.4 ms per loop (mean ± std. dev. of 7 runs, 1 loop each)        
                                                                           
shape: (1000, 1000), device: cpu, dtype: torch.float32         
9.42 s ± 14.7 ms per loop (mean ± std. dev. of 7 runs, 1 loop each)                                              
                                                                             
shape: (10, 1000, 1000), device: cpu, dtype: torch.float32                                                       
1min 50s ± 282 ms per loop (mean ± std. dev. of 7 runs, 1 loop each)

```
</details>

<details>

<summary>This PR, cpu float64</summary>

```
shape: (10, 10), device: cpu, dtype: torch.float64
381 µs ± 761 ns per loop (mean ± std. dev. of 7 runs, 1000 loops each)

shape: (1000, 10, 10), device: cpu, dtype: torch.float64
6.19 ms ± 13.2 µs per loop (mean ± std. dev. of 7 runs, 100 loops each)

shape: (100, 100), device: cpu, dtype: torch.float64
4.6 ms ± 3.26 µs per loop (mean ± std. dev. of 7 runs, 100 loops each)

shape: (1000, 100, 100), device: cpu, dtype: torch.float64
2.59 s ± 8.25 ms per loop (mean ± std. dev. of 7 runs, 1 loop each)

shape: (1000, 1000), device: cpu, dtype: torch.float64
1.07 s ± 5.09 ms per loop (mean ± std. dev. of 7 runs, 1 loop each)

shape: (10, 1000, 1000), device: cpu, dtype: torch.float64
14.4 s ± 13.9 ms per loop (mean ± std. dev. of 7 runs, 1 loop each)

```
</details>

<details>

<summary>Master, cpu float64</summary>

```
shape: (10, 10), device: cpu, dtype: torch.float64
395 µs ± 1.04 µs per loop (mean ± std. dev. of 7 runs, 1000 loops each)

shape: (1000, 10, 10), device: cpu, dtype: torch.float64
14.6 ms ± 9.76 µs per loop (mean ± std. dev. of 7 runs, 100 loops each)

shape: (100, 100), device: cpu, dtype: torch.float64
45.5 ms ± 154 µs per loop (mean ± std. dev. of 7 runs, 10 loops each)

shape: (1000, 100, 100), device: cpu, dtype: torch.float64
33.1 s ± 69.8 ms per loop (mean ± std. dev. of 7 runs, 1 loop each)

shape: (1000, 1000), device: cpu, dtype: torch.float64
19.3 s ± 80.3 ms per loop (mean ± std. dev. of 7 runs, 1 loop each)

shape: (10, 1000, 1000), device: cpu, dtype: torch.float64
3min 30s ± 1.29 s per loop (mean ± std. dev. of 7 runs, 1 loop each)

```
</details>

### CODE REVIEW

**Review by IvanYashchuk — APPROVED** [review]
The code looks good and speedups are great! Looking forward to the notes about the algorithmic details.

**Review by lezcano — COMMENTED** [review]
I've mostly read through the algorithm, and it was the one I had in mind. I left a few comments on the write-up, some things that could be explained more clearly. In particular, the point that we discussed about the fact that the current algorithm just works whenever `tau_i != 1/\norm{v_i}`.

I also left one comment regarding the readability of the implementation via a struct of `apply_left` and `apply_right` and two other comments on how to improve the efficiency by going from functions that use BLAS 3 operations to some that use BLAS 2 operations, which could potentially mean much faster execution times.

**Comment by nikitaved** [comment]
Thanks for your reviews, @Lezcano! I have addressed your comments. Please, have a look.

**Review by lezcano — APPROVED** [review]
Just left a few nits.

So, the only outstanding point is the use of matmul vs BLAS 2 when the inputs are vectors. I think this would be a good optimisation to have, given that this function is used on matrices in https://github.com/pytorch/pytorch/pull/62089

The only annoying point is that this would require having a duplicate path for virtually all the backward computation. Otherwise, you could open an Issue noting this, and offering as possible solutions using the function to compute the action of a HH-reflector on a matrix present on BLAS and the idea of using matrix-vector products when possible and then link this issue in the code as a TODO.

File: torch/csrc/autograd/FunctionsManual.cpp [hunk 1]
```diff
@@ -3430,104 +3430,169 @@ bool any_variable_defined(const variable_list& variables) {
...
+//
+// Note that K_{i + 1} = H_{i + 1}^{-1} K_i H_i, so we can compute v_i_grad, tau_i_grad one by one
+// by just efficiently updating K_i if that is possible. Multiplying with H_i from the right could be
+// done with matrix-vector products, but what about the inverse H_{i + 1}^{-1} and does it even exist?
+// Luckily, under some assumptions, H_{i + 1}^{-1} exists and admits a representation as H_i(sigma_i) for some
+// sigma_i, so the left update is also could be done with matrix-vector and not matrix-matrix products.
+//
+// Let H(tau) := I - tau v v^H.
+// H(tau) has eigenvalues 1 with multiplicity (m - 1) with eigenvectors orthogonal to v,
+// and an eigenvalue (1 - tau ||v||^2) with the corresponding eigenvector v / ||v||.
```
[the remark this answers is not in the document]
Comment (nikitaved) [reply torch/csrc/autograd/FunctionsManual.cpp:3462]: I prefer normalized to 1-norm vectors. If v is an eigenvectors so is t * v for t != 0.

File: torch/csrc/autograd/FunctionsManual.cpp [hunk 2]
```diff
@@ -3505,13 +3505,24 @@ std::tuple<Tensor, Tensor> householder_product_backward(const Tensor& grad, cons
...
   // With higher order derivatives we cannot rewrite the storage of K, hence we use much less efficient
   // out-of-place methods.
-  auto apply_householder_reflector = [&m](
-      int64_t k, const Tensor& v_full, const Tensor& t, Tensor& K, bool left = true, bool in_place = true) -> Tensor {
+  //
+  // if only first-order derivative is expected, we can modify K in-place for better performance
+  bool modify_K_in_place = !at::GradMode::is_enabled();
+
+  // TODO: replace this function with calls to LARFB
+  auto apply_householder_reflector = [&m, &modify_K_in_place](
```
Comment (Lezcano) [inline torch/csrc/autograd/FunctionsManual.cpp:3513]: Nit: As `modify_K_in_place` is a bool it's even better to capture it by value. Same for `m` as it's an integer.

Comment (nikitaved) [reply torch/csrc/autograd/FunctionsManual.cpp:3513]: Yeah, makes total sense, thanks!

**Comment by anjali411** [comment]
@nikitaved @Lezcano is this PR ready to be merged?

**Review by lezcano — APPROVED** [review]
LGTM

**Comment by mruberry** [comment]
@anjali411 are you merging this?

**Comment by nikitaved** [comment]
Gentle ping :)

**Review by anjali411 — APPROVED** [review]
